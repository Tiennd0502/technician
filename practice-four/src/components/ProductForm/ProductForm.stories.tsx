import type { Meta, StoryObj } from '@storybook/react';

import ProductForm from '.';

const meta: Meta<typeof ProductForm> = {
  title: 'Components/ProductForm',
  component: ProductForm,
};

export default meta;
type Story = StoryObj<typeof ProductForm>;

export const Default: Story = {
  args: {
    isOpen: true,
    onClose: () => null,
    onSubmit: () => null,
  },
};
