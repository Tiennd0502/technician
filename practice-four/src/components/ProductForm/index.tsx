import { memo } from 'react';
import {
  Button,
  Radio,
  Flex,
  Modal,
  ModalBody,
  ModalCloseButton,
  ModalContent,
  ModalFooter,
  ModalHeader,
  ModalOverlay,
  VStack,
  FormControl,
  FormLabel,
  RadioGroup,
} from '@chakra-ui/react';
import { Controller, SubmitHandler, useForm } from 'react-hook-form';

import { Product, STATUS } from '@/interfaces';
import { PRODUCT_RULES } from '@/constants';
import { Input } from '@/components';

interface ProductFormProps {
  title?: string;
  defaultValues?: Product;
  onSubmit: SubmitHandler<Product>;
  isSubmitting?: boolean;
  isOpen: boolean;
  onClose: VoidFunction;
}

const ProductForm = ({
  title = 'Add Product',
  defaultValues,
  onSubmit,
  isSubmitting,
  isOpen,
  onClose,
}: ProductFormProps) => {
  const { control, handleSubmit } = useForm<Product>({
    reValidateMode: 'onSubmit',
    defaultValues,
  });

  return (
    <Modal isOpen={isOpen} onClose={onClose}>
      <ModalOverlay />
      <ModalContent borderRadius='md' p='7' minW='90%'>
        <ModalHeader>{title}</ModalHeader>
        <ModalCloseButton />
        <ModalBody>
          <VStack as='form' onSubmit={handleSubmit(onSubmit)} id='form' spacing='2'>
            <Controller
              control={control}
              name='name'
              rules={PRODUCT_RULES.name}
              render={({ field: { value, ...field }, fieldState: { error } }) => (
                <Input
                  {...field}
                  error={error?.message || ''}
                  label='Product Name:'
                  defaultValue={value}
                  placeholder='Product Name'
                />
              )}
            />
            <Controller
              control={control}
              name='brand'
              rules={PRODUCT_RULES.service}
              render={({ field: { value, ...field }, fieldState: { error } }) => (
                <Input
                  {...field}
                  error={error?.message || ''}
                  label='Brand Name:'
                  defaultValue={value}
                  placeholder='Brand Name'
                />
              )}
            />
            <Controller
              control={control}
              name='service'
              rules={PRODUCT_RULES.service}
              render={({ field: { value, ...field }, fieldState: { error } }) => (
                <Input
                  {...field}
                  error={error?.message || ''}
                  label='Service:'
                  defaultValue={value}
                  placeholder='Service'
                />
              )}
            />
            <Controller
              control={control}
              name='status'
              render={({ field: { value, ...field } }) => (
                <FormControl>
                  <FormLabel>Status:</FormLabel>
                  <RadioGroup
                    defaultValue={value?.toString() || STATUS.Deactivated.toString()}
                    {...field}
                  >
                    <Flex gap='3'>
                      <Radio {...field} value={STATUS.Activated.toString()}>
                        Activated
                      </Radio>
                      <Radio value={STATUS.Deactivated.toString()}>Deactivated</Radio>
                    </Flex>
                  </RadioGroup>
                </FormControl>
              )}
            />
          </VStack>
        </ModalBody>
        <ModalFooter as={Flex} justifyContent='center'>
          <Button variant='outline' isDisabled={isSubmitting} ml='auto' mr='6' onClick={onClose}>
            Cancel
          </Button>
          <Button type='submit' form='form' isLoading={isSubmitting}>
            Submit
          </Button>
        </ModalFooter>
      </ModalContent>
    </Modal>
  );
};

export default memo(ProductForm);
