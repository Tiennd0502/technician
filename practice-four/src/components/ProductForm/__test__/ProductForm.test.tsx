import { act, fireEvent, render, screen, waitFor } from '@testing-library/react';

import ProductForm from '..';

const onSubmit = jest.fn();
const onClose = jest.fn();

describe('ProductForm component', () => {
  const props = {
    isOpen: true,
    onClose,
    onSubmit,
  };

  test('should show error message when form got error', async () => {
    const { getByRole } = render(<ProductForm {...props} />);
    const submit = getByRole('button', { name: /Submit/i });
    fireEvent.click(submit);

    await waitFor(() => {
      const requireMessage = screen.getByText('Product Name is required');

      expect(requireMessage).toBeTruthy();
    });
  });

  test('should submit form', async () => {
    const { getByPlaceholderText, getByRole } = render(<ProductForm {...props} />);

    const nameInput = getByPlaceholderText('Product Name');
    const brandInput = getByPlaceholderText('Brand Name');
    const serviceInput = getByPlaceholderText('Service');
    const submit = getByRole('button', { name: /Submit/i });

    await act(async () => {
      fireEvent.change(nameInput, { target: { value: 'Product 1' } });
      fireEvent.change(brandInput, { target: { value: 'LG' } });
      fireEvent.change(serviceInput, { target: { value: 'Repair' } });
      fireEvent.click(submit);
    });

    expect(onSubmit).toHaveBeenCalledTimes(1);
  });
});
