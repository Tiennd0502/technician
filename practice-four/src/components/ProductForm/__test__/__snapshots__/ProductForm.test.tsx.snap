// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductForm component it matches the snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div
      aria-hidden="true"
      data-aria-hidden="true"
    />
    <div
      class="chakra-portal"
    >
      <div
        aria-hidden="true"
        class="chakra-modal__overlay css-13d1tkw"
        data-aria-hidden="true"
        style="opacity: 0;"
      />
      <div
        aria-hidden="true"
        data-aria-hidden="true"
        data-focus-guard="true"
        style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
        tabindex="0"
      />
      <div
        data-focus-lock-disabled="false"
      >
        <div
          class="chakra-modal__content-container css-1oxhx2p"
          tabindex="-1"
        >
          <section
            aria-describedby="chakra-modal--body-:rk:"
            aria-labelledby="chakra-modal--header-:rk:"
            aria-modal="true"
            class="chakra-modal__content css-17pp93c"
            id="chakra-modal-:rk:"
            role="dialog"
            style="opacity: 0; transform: scale(0.95) translateZ(0);"
            tabindex="-1"
          >
            <header
              class="chakra-modal__header css-72fd9l"
              id="chakra-modal--header-:rk:"
            >
              Add Product
            </header>
            <button
              aria-label="Close"
              class="chakra-modal__close-btn css-6sdaxy"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="chakra-icon css-onkibi"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M.439,21.44a1.5,1.5,0,0,0,2.122,2.121L11.823,14.3a.25.25,0,0,1,.354,0l9.262,9.263a1.5,1.5,0,1,0,2.122-2.121L14.3,12.177a.25.25,0,0,1,0-.354l9.263-9.262A1.5,1.5,0,0,0,21.439.44L12.177,9.7a.25.25,0,0,1-.354,0L2.561.44A1.5,1.5,0,0,0,.439,2.561L9.7,11.823a.25.25,0,0,1,0,.354Z"
                  fill="currentColor"
                />
              </svg>
            </button>
            <div
              class="chakra-modal__body css-0"
              id="chakra-modal--body-:rk:"
            >
              <form
                class="chakra-stack css-11nd7jy"
                id="form"
              >
                <div
                  class="chakra-form-control css-0"
                  role="group"
                >
                  <label
                    class="chakra-form__label css-gu124"
                    for="field-:rn:"
                    id="field-:rn:-label"
                  >
                    Product Name:
                  </label>
                  <input
                    class="chakra-input css-jgxprw"
                    id="field-:rn:"
                    name="name"
                    placeholder="Product Name"
                    type="text"
                    value=""
                  />
                </div>
                <div
                  class="chakra-form-control css-0"
                  role="group"
                >
                  <label
                    class="chakra-form__label css-gu124"
                    for="field-:ro:"
                    id="field-:ro:-label"
                  >
                    Brand Name:
                  </label>
                  <input
                    class="chakra-input css-jgxprw"
                    id="field-:ro:"
                    name="brand"
                    placeholder="Brand Name"
                    type="text"
                    value=""
                  />
                </div>
                <div
                  class="chakra-form-control css-0"
                  role="group"
                >
                  <label
                    class="chakra-form__label css-gu124"
                    for="field-:rp:"
                    id="field-:rp:-label"
                  >
                    Service:
                  </label>
                  <input
                    class="chakra-input css-jgxprw"
                    id="field-:rp:"
                    name="service"
                    placeholder="Service"
                    type="text"
                    value=""
                  />
                </div>
                <div
                  class="chakra-form-control css-0"
                  role="group"
                >
                  <label
                    class="chakra-form__label css-uvho5g"
                    for="field-:rq:"
                    id="field-:rq:-label"
                  >
                    Status:
                  </label>
                  <div
                    class="chakra-radio-group css-0"
                    role="radiogroup"
                  >
                    <div
                      class="css-blvggq"
                    >
                      <label
                        class="chakra-radio css-1nqw95c"
                      >
                        <input
                          class="chakra-radio__input"
                          id="radio-:rs:"
                          name="status"
                          style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                          type="radio"
                          value="1"
                        />
                        <span
                          aria-hidden="true"
                          class="chakra-radio__control css-ssalds"
                        />
                        <span
                          class="chakra-radio__label css-1y8kf23"
                        >
                          Activated
                        </span>
                      </label>
                      <label
                        class="chakra-radio css-1nqw95c"
                        data-checked=""
                      >
                        <input
                          checked=""
                          class="chakra-radio__input"
                          id="radio-:rt:"
                          name="status"
                          style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                          type="radio"
                          value="0"
                        />
                        <span
                          aria-hidden="true"
                          class="chakra-radio__control css-ssalds"
                          data-checked=""
                        />
                        <span
                          class="chakra-radio__label css-1y8kf23"
                          data-checked=""
                        >
                          Deactivated
                        </span>
                      </label>
                    </div>
                  </div>
                </div>
              </form>
            </div>
            <div
              class="chakra-modal__footer css-1ygmb7j"
            >
              <button
                class="chakra-button css-oglbjv"
                type="button"
              >
                Cancel
              </button>
              <button
                class="chakra-button css-4xx2wk"
                form="form"
                type="submit"
              >
                Submit
              </button>
            </div>
          </section>
        </div>
      </div>
      <div
        aria-hidden="true"
        data-aria-hidden="true"
        data-focus-guard="true"
        style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
        tabindex="0"
      />
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
    data-aria-hidden="true"
  />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
